### PR TITLE
Add API create event publishing to gateway

### PR DESCRIPTION
**Pull request by tgtshanika**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    boolean publishAPICreateEventToGateway(API api) throws GatewayException;
+
+    /**
+     * Publish API update event to gateway
+     *
+     * @param api
+     * @return
+     * @throws GatewayException
+     */
+    boolean publishAPIUpdateEventToGateway(API api) throws GatewayException;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java]: please add description related to the params etc

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Publish API delete event to gateway
+     *
+     * @param api
+     * @return
+     * @throws GatewayException
+     */
+
+    boolean publishAPIDeleteEventToGateway(API api) throws GatewayException;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java]: same here

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     *
+     *
+     * @param endpointConfig
+     * @return
+     * @throws GatewayException
+     */
+    boolean publishEndpointCreateEventToGateway(String endpointConfig) throws GatewayException;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java]: same here. also missing discription

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java [hunk 4]
```diff
@@ -47,27 +50,24 @@
 /**
  * API gateway related functions
  */
-public class APIGatewayPublisherImpl implements APIGatewayPublisher {
-    private static final Logger log = LoggerFactory.getLogger(APIGatewayPublisherImpl.class);
+public class APIGatewayEventPublisherImpl implements APIGatewayEventPublisher {
+    private static final Logger log = LoggerFactory.getLogger(APIGatewayEventPublisherImpl.class);
     private APIMConfigurations config;
     private String gatewayFileExtension = ".bal";
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java]: set these to final

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     * @param api                   API definition
+     * @return                      Is published successfully
+     * @throws GatewayException     If there is a failure of publishing
+     */
+    boolean publishAPICreateEventToGateway(API api) throws GatewayException;
+
+    /**
+     * Publish API update event to gateway
+     *
+     * @param api
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGatewayEventPublisher.java]: JavaDoc incomplete. Better to provide more details

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java [hunk 6]
```diff
@@ -47,27 +50,24 @@
...
     private String endpointConfigName = "endpoint";
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java]: same here

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java [hunk 7]
```diff
@@ -84,18 +84,12 @@ public boolean publishToGateway(API api) throws GatewayException {
...
-                dto.setConfig(gatewayConfig);
-                publishMessage(dto); //publishing versioned API
-                if (api.isDefaultVersion()) {
-                    //publishing default API
-                    publishMessage(dto);
-                }
+                APIGatewayEvent apiCreateEvent = new APIGatewayEvent(APIMgtConstants.
+                        APIGatewayEventTypes.API_GW_EVENT_TYPE_API_CREATE);
+                apiCreateEvent.setGatewayLabels(api.getLabels());
+                apiCreateEvent.addEventDetail("apiId", api.getId());
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java]: Shall we move "apiId" for a constant

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java [hunk 8]
```diff
@@ -205,13 +214,10 @@ private void saveApi(API api, String gwHome, String content, boolean isDefaultAp
...
-     *
-     * @param config endpoint config
-     * @return is publishing success
+     * @see APIGatewayEventPublisher#publishEndpointCreateEventToGateway(String)
      */
     @Override
-    public boolean publishEndpointConfigToGateway(String config) throws GatewayException {
+    public boolean publishEndpointCreateEventToGateway(String endpointConfig) throws GatewayException {
         try {
             String gwHome = System.getProperty("gwHome");
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java]: add a constant

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/models/APIGatewayEvent.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        this.eventType = eventType;
+    }
+
+    public String getEventDetail(String key) {
+        return eventDetails.get(key);
+    }
+
+    public void addEventDetail(String key, String value) {
+        eventDetails.put(key, value);
+    }
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/models/APIGatewayEvent.java]: shall we add a toString() method and then use it later for debug logs. would be easier to identify issues later

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java [hunk 10]
```diff
@@ -221,13 +227,12 @@ public boolean publishEndpointConfigToGateway(String config) throws GatewayExcep
...
             if (gwHome == null) {
-                GatewayConfigDTO dto = new GatewayConfigDTO();
-                dto.setType("endpoint");
-                dto.setApiName(endpointConfigName);
-                dto.setConfig(config);
-                publishMessage(dto); //TODO publish endpoint configs correctly
+                APIGatewayEvent endpointCreateEvent = new APIGatewayEvent(APIMgtConstants.
+                        APIGatewayEventTypes.API_GW_EVENT_TYPE_ENDPOINT_CREATE);
+
+                publishMessage(endpointCreateEvent, config.getTopicName()); //TODO publish endpoint configs correctly
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayEventPublisherImpl.java]: Is this todo completed or it is still remaining task?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java [hunk 11]
```diff
@@ -320,8 +321,11 @@ public String addAPI(API.APIBuilder apiBuilder) throws APIManagementException {
                         getApiDAO().addAPI(createdAPI);
                     }
                 }
+
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: Remove the unwanted blank lines

**Pull request closed without merging**